### PR TITLE
feat(SPV-1061) new EF converter

### PR DESCRIPTION
**Pull request by chris-4chain**
Another piece for new future broadcaster. 

## AI generated summary
This pull request introduces a new `Converter` class for converting transactions to EF format and includes comprehensive unit tests and error handling. The most important changes are the creation of the `Converter` class, the addition of error handling, and the implementation of unit tests.

### New Features:
* [`engine/chain/internal/ef/converter.go`](diffhunk://#diff-d61bb1f0b528974b287f960a7a6e3ecf1774aa4c74ee195bb10623ac2846820aR1-R66): Added a `Converter` class with a method `Convert` to convert transactions to EF format, including handling unsourced inputs and hydrating them with source transactions.

### Error Handling:
* [`engine/chain/internal/ef/errors.go`](diffhunk://#diff-70b457e5ec06269c01e3aa870829158a68d927e0263a937481f21d50878dd059R1-R9): Introduced new error types `ErrMissingSourceTXID`, `ErrGetTransactions`, and `ErrEFHexGeneration` for better error handling in the conversion process.

### Unit Tests:
* [`engine/chain/internal/ef/converter_test.go`](diffhunk://#diff-a4c97a654e8b84b75f313562df73224f1e1518464efaf1b34b436fcbf2af0a3cR1-R139): Added unit tests for the `Converter` class, covering various scenarios including successful conversions and different error cases.
* [`engine/chain/internal/ef/mock_txgetter_test.go`](diffhunk://#diff-f6f8b6d6174787bae03ccef1eeeca52adce3a544ebf382ff33fca0ad9c781aa6R1-R77): Implemented a mock `TransactionsGetter` for testing purposes, including different behaviors for missing transactions.

### Helper Classes:
* [`engine/chain/internal/ef/unsourced_inputs.go`](diffhunk://#diff-d6779e4f298f3b59348068dde3e8c21c4da819cbe5d1732b13e3eae682f4d71aR1-R64): Added `unsourcedInputsManager` to manage unsourced inputs in transactions, including methods to find, count, and hydrate unsourced inputs.

# Pull Request Checklist

- [ ] 📖 I created my PR using provided  : [CODE_STANDARDS](https://github.com/bitcoin-sv/spv-wallet/blob/main/.github/CODE_STANDARDS.md)
- [ ] 📖 I have read the short Code of Conduct: [CODE_OF_CONDUCT](https://github.com/bitcoin-sv/spv-wallet/blob/main/.github/CODE_OF_CONDUCT.md)
- [ ] 🏠 I tested my changes locally.
- [ ] ✅ I have provided tests for my changes.
- [ ] 📝 I have used conventional commits.
- [ ] 📗 I have updated any related documentation.
- [ ] 💾 PR was issued based on the Github or Jira issue.

### CODE REVIEW

File: engine/chain/internal/ef/converter_test.go [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+			require.ErrorIs(t, err, test.expectErr)
+			require.Empty(t, efHex)
+		})
+	}
+}
+
+func txFromBase64(t *testing.T, rawTXBase64 string) *sdk.Transaction {
+	d, err := base64.StdEncoding.DecodeString(rawTXBase64)
+	require.NoError(t, err)
+	tx, err := sdk.NewTransactionFromBytes(d)
```
Comment (dorzepowski) [inline engine/chain/internal/ef/converter_test.go:137]: NewTransactionFromHex wouldn't work here?

File: engine/chain/internal/ef/converter_test.go [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+			require.Equal(t, test.expectedEFHex, efHex)
+
+			// additionally check if converter can convert already-in-ef-hex tx
+			// without the need to fetch source transactions
+			tx, err = sdk.NewTransactionFromHex(efHex)
+			require.NoError(t, err)
+			converter = ef.NewConverter(newMockTransactionsGetter(t, []string{}))
+			efHexRegenerated, err := converter.Convert(context.Background(), tx)
+			require.NoError(t, err)
+			require.Equal(t, efHex, efHexRegenerated)
```
Comment (dorzepowski) [inline engine/chain/internal/ef/converter_test.go:73]: for me it should be a single separated test case, to make it clear why it's failing.

File: engine/chain/internal/ef/converter_test.go [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+	"testing"
+
+	sdk "github.com/bitcoin-sv/go-sdk/transaction"
+	"github.com/bitcoin-sv/spv-wallet/engine/chain/internal/ef"
+	"github.com/stretchr/testify/require"
+)
+
+const (
+	// https://whatsonchain.com/tx/2978f03c8a21bf90b5980113f988c39ef4ae691b9bedd5178c50ebb9c034dabf
+	tx1       = "AQAAAAF3b50sDYC2EspU68ofo7042zdXVux3eO3dwyNWngbclgEAAABrSDBFAiEA+DQVdQiAzJRkt1LDMhXt51aMRcg808y4QXh+3RIZ02gCIGXnGwkkHIiVKeG5ecnMn1aSYzhoEfAAVIWyYXE0ci3QQSECDQrOYn+/gOIM5U2Lz6WqQfSm8tnBE7rAzxsTFqlvDJ//////AgEAAAAAAAAAGXapFH1x4vjOGUCfuTMSyfAZp97j0U6hiKwMAAAAAAAAABl2qRTlpPBsGiTCqd3/yuRAWq9sIDuj0oisAAAAAA=="
```
Comment (dorzepowski) [inline engine/chain/internal/ef/converter_test.go:15]: can you name it more detailed, like: transactionWithSingleInput

File: engine/chain/internal/ef/converter_test.go [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+	tx1Source = "AQAAAAH3r1bx2GGVS0wrWlVhnlfTzPJmpx754/9+Wt/v2PhdMAEAAABqRzBEAiAKny96EiIym5fiO8XE3vCCNgBJXir6O8aue+XVu9av8gIgUl9pJF9SVIBmxTGUI1eGiO9GurGLgYVkXb4lh/DJ27NBIQI3r2kZnSB/9c4jCWl2ROrCur6oPq3v2NODhMjK4+h83/////8CAQAAAAAAAAAZdqkUb96slVk3zuwEk+aWBCGEljZyf8CIrA4AAAAAAAAAGXapFJyF4WV/sj1B4UcSSt7233kz6GyyiKwAAAAA"
+
+	// https://whatsonchain.com/tx/88a7c0ed1cb4767cfc8e7434561379eaea21ae78e480cacf4e69284387057c70
+	tx2        = "AQAAAAIbSuUDkTFyxeFr2J2rtx01PFucsqHGmXD9TmkOSfl0EAEAAABqRzBEAiAxJ9U+0u2IQ9la0NpJZZ4IbimNyMSr+UZlburh/VyMhgIgXhDzvSw/AcCJA8OWnRONYrHnbJb4VuJ0PPMGnLRpWnVBIQJ5Ili3+6UMih1hVPC0vkpOV7B47+G0eUbAEGl+md3nkf/////IjkyHDWHX4UuJMdlByIj/s2rVjFI2TknC3yD1ZdrezQEAAABrSDBFAiEAxCUx8LUKyrb9H2OzCisQRqwpllvAz0FAmxgNPUuRq+wCIAb7rabUlp3l8WKX+JBaigB2PKGaGiesMbHnfRKHd6FAQSED0h5ymG3g01Sv8d1zegZra3hrwgS+wis5QeEOlXWnqnv/////AhQAAAAAAAAAGXapFOiWQpj8qlBvOebR0fKWV/ecHnLniKwJAAAAAAAAABl2qRTgvT8tXBkZEJgxv61AuOspPAdiG4isAAAAAA=="
```
Comment (dorzepowski) [inline engine/chain/internal/ef/converter_test.go:19]: can you name it more detailed, like: transactionWithMultipleInputs

**Review by dorzepowski — CHANGES_REQUESTED** [review]

File: engine/chain/internal/ef/converter_test.go (hunk 1 above)
Comment (chris-4chain) [reply engine/chain/internal/ef/converter_test.go:137]: I've manually converted them from base64 to hex. 
So, as you say, `NewTransactionFromHex` works.

FYI: I have loaded the rawTX using junglenus Json API. And there are base64 encoded strings.

**Review by dorzepowski — APPROVED** [review]

**Pull request merged**